mpz-core/src/lpn.rs

```diff
@@ -0,0 +1,156 @@
+//! Implement LPN with local linear code.
+//! More especifically, a local linear code is a random boolean matrix with at most D non-zero values in each row.
+
+use crate::{prp::Prp, Block};
+use rayon::prelude::*;
+/// A struct related to LPN.
+/// The `seed` defines a sparse binary matrix `A` with at most `D` non-zero values in each row.
+/// `A` - is a binary matrix with `k` columns and `n` rows. The concrete number of `n` is determined by the input length. `A` will be generated on-the-fly.
+/// `x` - is a `F_{2^128}` vector with length `k`.
+/// `e` - is a `F_{2^128}` vector with length `n`.
+/// Given a vector `x` and `e`, compute `y = Ax + e`.
+/// Note that in the standard LPN problem, `x` is a binary vector, `e` is a sparse binary vector. The way we difined here is a more generic way in term of computing `y`.
+pub struct Lpn<const D: usize> {
+    // The seed to generate the random sparse matrix A.
+    seed: Block,
+
+    // The length of the secret, i.e., x.
+    k: u32,
+
+    // A mask to optimize reduction operation.
+    mask: u32,
+}
+
+impl<const D: usize> Lpn<D> {
+    /// New an LPN instance
+    pub fn new(seed: Block, k: u32) -> Self {
+        let mut mask = 1;
+        while mask < k {
+            mask <<= 1;
+            mask |= 0x1;
+        }
+        Self { seed, k, mask }
+    }
+
+    // Compute 4 rows as a batch, this is for the `compute_naive` function.
+    #[inline]
+    fn compute_four_rows_non_indep(&self, y: &mut [Block], x: &[Block], pos: usize, prp: &Prp) {
+        let mut cnt = 0u64;
+        let index = [0; D].map(|_| {
+            let i: u64 = cnt;
+            cnt += 1;
+            Block::from(bytemuck::cast::<_, [u8; 16]>([pos as u64, i]))
+        });
+
+        let mut index: [Block; D] = prp.permute_many_blocks(index);
+        let index: &mut [u32] = bytemuck::cast_slice_mut::<_, u32>(&mut index);
+
+        for (i, y) in y[pos..].iter_mut().enumerate().take(4) {
+            for ind in index[i * D..(i + 1) * D].iter_mut() {
+                *ind &= self.mask;
+                *ind = if *ind >= self.k { *ind - self.k } else { *ind };
+
+                *y ^= x[*ind as usize];
+            }
+        }
+    }
+
+    // Compute 4 rows as a batch, this is for the `compute` function.
+    #[inline]
+    fn compute_four_rows_indep(&self, y: &mut [Block], x: &[Block], pos: usize, prp: &Prp) {
+        let mut cnt = 0u64;
+        let index = [0; D].map(|_| {
+            let i = cnt;
+            cnt += 1;
+            Block::from(bytemuck::cast::<_, [u8; 16]>([pos as u64, i]))
+        });
```

Comment:
`std::array::from_fn` is handy here:

```suggestion
        let index: [Block; D] = std::array::from_fn(|_| {
            let i = cnt;
            cnt += 1;
            Block::from(bytemuck::cast::<_, [u8; 16]>([pos as u64, i]))
        });
```